File: management-api/src/routes/opencode.ts
/**
 * OpenCode Proxy Routes
 * 
 * Proxies requests to OpenCode containers running on Coolify.
 * All routes are under /api/projects/:id/opencode/*
 */

import { Hono } from 'hono';
import { opencode, OpenCodeProxyError } from '../services/opencode.ts';
import { getProjectById } from '../models/project.ts';
import { coolify } from '../services/coolify.ts';
import { config } from '../config.ts';
import { createLogger } from '../utils/logger.ts';

const log = createLogger('opencode-routes');

export const opencodeRoutes = new Hono()

  // ===========================================================================
  // App Info / Health
  // ===========================================================================

  /**
   * GET /api/projects/:id/opencode/app
   * Get OpenCode app info (health check)
   */
  .get('/:id/opencode/app', async (c) => {
    const projectId = c.req.param('id');
    
    try {
      const info = await opencode.getAppInfo(projectId);
      return c.json(info);
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to get OpenCode app info', { projectId, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  // ===========================================================================
  // Sessions
  // ===========================================================================

  /**
   * GET /api/projects/:id/opencode/session
   * List all sessions
   */
  .get('/:id/opencode/session', async (c) => {
    const projectId = c.req.param('id');
    
    try {
      const sessions = await opencode.listSessions(projectId);
      return c.json(sessions);
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to list sessions', { projectId, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  /**
   * POST /api/projects/:id/opencode/session
   * Create a new session
   */
  .post('/:id/opencode/session', async (c) => {
    const projectId = c.req.param('id');
    
    try {
      const session = await opencode.createSession(projectId);
      return c.json(session, 201);
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to create session', { projectId, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  /**
   * GET /api/projects/:id/opencode/session/:sessionId
   * Get session details
   */
  .get('/:id/opencode/session/:sessionId', async (c) => {
    const projectId = c.req.param('id');
    const sessionId = c.req.param('sessionId');
    
    try {
      const session = await opencode.getSession(projectId, sessionId);
      return c.json(session);
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to get session', { projectId, sessionId, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  /**
   * DELETE /api/projects/:id/opencode/session/:sessionId
   * Delete a session
   */
  .delete('/:id/opencode/session/:sessionId', async (c) => {
    const projectId = c.req.param('id');
    const sessionId = c.req.param('sessionId');
    
    try {
      await opencode.deleteSession(projectId, sessionId);
      return c.json({ success: true });
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to delete session', { projectId, sessionId, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  /**
   * POST /api/projects/:id/opencode/session/:sessionId/abort
   * Abort a running session
   */
  .post('/:id/opencode/session/:sessionId/abort', async (c) => {
    const projectId = c.req.param('id');
    const sessionId = c.req.param('sessionId');
    
    try {
      await opencode.abortSession(projectId, sessionId);
      return c.json({ success: true });
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to abort session', { projectId, sessionId, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  // ===========================================================================
  // Messages
  // ===========================================================================

  /**
   * GET /api/projects/:id/opencode/session/:sessionId/message
   * List messages in a session
   */
  .get('/:id/opencode/session/:sessionId/message', async (c) => {
    const projectId = c.req.param('id');
    const sessionId = c.req.param('sessionId');
    
    try {
      const messages = await opencode.listMessages(projectId, sessionId);
      return c.json(messages);
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to list messages', { projectId, sessionId, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  /**
   * POST /api/projects/:id/opencode/session/:sessionId/message
   * Send a message to a session
   */
  .post('/:id/opencode/session/:sessionId/message', async (c) => {
    const projectId = c.req.param('id');
    const sessionId = c.req.param('sessionId');
    
    try {
      const body = await c.req.json();
      const message = await opencode.sendMessage(projectId, sessionId, body);
      return c.json(message, 201);
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to send message', { projectId, sessionId, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  /**
   * GET /api/projects/:id/opencode/session/:sessionId/message/:messageId
   * Get a specific message
   */
  .get('/:id/opencode/session/:sessionId/message/:messageId', async (c) => {
    const projectId = c.req.param('id');
    const sessionId = c.req.param('sessionId');
    const messageId = c.req.param('messageId');
    
    try {
      const message = await opencode.getMessage(projectId, sessionId, messageId);
      return c.json(message);
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to get message', { projectId, sessionId, messageId, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  // ===========================================================================
  // Files
  // ===========================================================================

  /**
   * GET /api/projects/:id/opencode/file
   * List files in the project
   * Query param: path (optional, defaults to "/")
   */
  .get('/:id/opencode/file', async (c) => {
    const projectId = c.req.param('id');
    const path = c.req.query('path') || '/';
    
    try {
      const files = await opencode.listFiles(projectId, path);
      return c.json(files);
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to list files', { projectId, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  /**
   * GET /api/projects/:id/opencode/file/content
   * Get file content
   * Query param: path (required)
   */
  .get('/:id/opencode/file/content', async (c) => {
    const projectId = c.req.param('id');
    const filePath = c.req.query('path');
    
    if (!filePath) {
      return c.json({ error: 'Missing required query parameter: path' }, 400);
    }
    
    try {
      const content = await opencode.getFileContent(projectId, filePath);
      return c.json(content);
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to get file content', { projectId, filePath, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  /**
   * GET /api/projects/:id/opencode/find/file
   * Find files by query
   * Query param: query (required)
   */
  .get('/:id/opencode/find/file', async (c) => {
    const projectId = c.req.param('id');
    const query = c.req.query('query');
    
    if (!query) {
      return c.json({ error: 'Missing required query parameter: query' }, 400);
    }
    
    try {
      const files = await opencode.findFiles(projectId, query);
      return c.json(files);
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to find files', { projectId, query, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  // ===========================================================================
  // SSE Events (Streaming)
  // ===========================================================================

  /**
   * GET /api/projects/:id/opencode/event
   * Proxy SSE event stream from OpenCode container
   * 
   * This endpoint streams Server-Sent Events from the OpenCode container.
   * Events include: session.updated, message.part.updated, tool.execute, etc.
   * 
   * Uses the OpenCode SDK to subscribe to events and forwards them to the client.
   */
  .get('/:id/opencode/event', async (c) => {
    const projectId = c.req.param('id');
    
    try {
      log.info('Starting SSE proxy via SDK', { projectId });
      
      // Get the async iterator of events from the SDK
      const eventIterator = await opencode.subscribeToEvents(projectId);
      const encoder = new TextEncoder();
      
      // Create a ReadableStream that forwards SDK events as SSE
      const stream = new ReadableStream({
        start(controller) {
          // Send initial connection comment immediately
          controller.enqueue(encoder.encode(': connected\n\n'));
          
          // Spawn async loop to process events (don't await - this runs in background)
          (async () => {
            try {
              // Iterate over SDK events and forward them
              for await (const event of eventIterator) {
                const eventType = event.type || 'message';
                const eventData = JSON.stringify(event);
                const sseMessage = `event: ${eventType}\ndata: ${eventData}\n\n`;
                controller.enqueue(encoder.encode(sseMessage));
              }
            } catch (err) {
              log.error('SSE stream error', { projectId, error: err });
              // Send error event before closing
              const errorEvent = `event: error\ndata: ${JSON.stringify({ message: 'Stream error' })}\n\n`;
              controller.enqueue(encoder.encode(errorEvent));
            } finally {
              controller.close();
            }
          })();
        },
        cancel() {
          log.info('SSE client disconnected', { projectId });
          // The async iterator will be garbage collected
        }
      });
      
      log.info('SSE connection established via SDK', { projectId });
      
      return new Response(stream, {
        headers: {
          'Content-Type': 'text/event-stream',
          'Cache-Control': 'no-cache',
          'Connection': 'keep-alive',
          'X-Accel-Buffering': 'no', // Disable nginx buffering
        },
      });
      
    } catch (error) {
      if (error instanceof OpenCodeProxyError) {
        return c.json({ error: error.message, details: error.details }, error.statusCode as 400);
      }
      log.error('Failed to start SSE proxy', { projectId, error });
      return c.json({ error: 'Internal server error' }, 500);
    }
  })

  // ===========================================================================
  // Health Check
  // ===========================================================================

  /**
   * GET /api/projects/:id/opencode/health
   * Check if OpenCode container is reachable
   */
  .get('/:id/opencode/health', async (c) => {
    const projectId = c.req.param('id');
    
    try {
      const healthy = await opencode.healthCheck(projectId);
      return c.json({ healthy, projectId });
    } catch (error) {
      return c.json({ healthy: false, projectId, error: error instanceof Error ? error.message : 'Unknown error' });
    }
  });

File: src-tauri/src/commands/opencode.rs
//! OpenCode commands
//!
//! Tauri commands for interacting with OpenCode containers via the Management API.
//! All communication goes through the Management API proxy.

use crate::models::{
    AppError, AppInfo, FileContent, FileNode, Message, MessagePartInput,
    OpenCodeEvent, OpenCodeHealth, SendMessageInput, Session, StreamConnection,
    StreamEventPayload, StreamStatus, StreamStatusPayload,
};
use crate::services::ApiClient;
use futures::StreamExt;
use std::collections::HashMap;
use std::sync::Arc;
use tauri::{AppHandle, Emitter, Manager};
use tokio::sync::Mutex;

/// Helper to get an authenticated API client
fn get_client() -> Result<ApiClient, AppError> {
    ApiClient::new()
}

// =============================================================================
// App Info & Health
// =============================================================================

/// Get OpenCode app info for a project
#[tauri::command]
pub async fn opencode_get_app_info(project_id: String) -> Result<AppInfo, AppError> {
    let client = get_client()?;
    client.opencode_get_app_info(&project_id).await
}

/// Check if OpenCode container is healthy
#[tauri::command]
pub async fn opencode_health_check(project_id: String) -> Result<OpenCodeHealth, AppError> {
    let client = get_client()?;
    client.opencode_health_check(&project_id).await
}

// =============================================================================
// Sessions
// =============================================================================

/// List all sessions for a project
#[tauri::command]
pub async fn opencode_list_sessions(project_id: String) -> Result<Vec<Session>, AppError> {
    let client = get_client()?;
    client.opencode_list_sessions(&project_id).await
}

/// Create a new session for a project
#[tauri::command]
pub async fn opencode_create_session(project_id: String) -> Result<Session, AppError> {
    let client = get_client()?;
    client.opencode_create_session(&project_id).await
}

/// Get a session by ID
#[tauri::command]
pub async fn opencode_get_session(
    project_id: String,
    session_id: String,
) -> Result<Session, AppError> {
    let client = get_client()?;
    client.opencode_get_session(&project_id, &session_id).await
}

/// Delete a session
#[tauri::command]
pub async fn opencode_delete_session(project_id: String, session_id: String) -> Result<(), AppError> {
    let client = get_client()?;
    client
        .opencode_delete_session(&project_id, &session_id)
        .await
}

/// Abort a running session
#[tauri::command]
pub async fn opencode_abort_session(project_id: String, session_id: String) -> Result<(), AppError> {
    let client = get_client()?;
    client
        .opencode_abort_session(&project_id, &session_id)
        .await
}

// =============================================================================
// Messages
// =============================================================================

/// List messages in a session
#[tauri::command]
pub async fn opencode_list_messages(
    project_id: String,
    session_id: String,
) -> Result<Vec<Message>, AppError> {
    let client = get_client()?;
    client
        .opencode_list_messages(&project_id, &session_id)
        .await
}

/// Send a text message to a session
#[tauri::command]
pub async fn opencode_send_message(
    project_id: String,
    session_id: String,
    text: String,
) -> Result<Message, AppError> {
    let client = get_client()?;

    let input = SendMessageInput {
        parts: vec![MessagePartInput {
            part_type: "text".to_string(),
            text: Some(text),
            url: None,
            filename: None,
            mime: None,
        }],
    };

    client
        .opencode_send_message(&project_id, &session_id, input)
        .await
}

/// Send a message with file attachments
#[tauri::command]
pub async fn opencode_send_message_with_files(
    project_id: String,
    session_id: String,
    text: String,
    files: Vec<String>, // File paths
) -> Result<Message, AppError> {
    let client = get_client()?;

    let mut parts = vec![MessagePartInput {
        part_type: "text".to_string(),
        text: Some(text),
        url: None,
        filename: None,
        mime: None,
    }];

    // Add file parts
    for file_path in files {
        let filename = file_path
            .split('/')
            .last()
            .unwrap_or(&file_path)
            .to_string();

        parts.push(MessagePartInput {
            part_type: "file".to_string(),
            text: None,
            url: Some(format!("file://{}", file_path)),
            filename: Some(filename),
            mime: Some("text/plain".to_string()),
        });
    }

    let input = SendMessageInput { parts };

    client
        .opencode_send_message(&project_id, &session_id, input)
        .await
}

/// Get a specific message
#[tauri::command]
pub async fn opencode_get_message(
    project_id: String,
    session_id: String,
    message_id: String,
) -> Result<Message, AppError> {
    let client = get_client()?;
    client
        .opencode_get_message(&project_id, &session_id, &message_id)
        .await
}

// =============================================================================
// Files
// =============================================================================

/// List files in a project directory
#[tauri::command]
pub async fn opencode_list_files(project_id: String, path: Option<String>) -> Result<Vec<FileNode>, AppError> {
    let client = get_client()?;
    let path = path.unwrap_or_else(|| "/".to_string());
    client.opencode_list_files(&project_id, &path).await
}

/// Get file content
#[tauri::command]
pub async fn opencode_get_file_content(
    project_id: String,
    path: String,
) -> Result<FileContent, AppError> {
    let client = get_client()?;
    client.opencode_get_file_content(&project_id, &path).await
}

/// Find files by pattern
#[tauri::command]
pub async fn opencode_find_files(
    project_id: String,
    pattern: String,
) -> Result<Vec<String>, AppError> {
    let client = get_client()?;
    client.opencode_find_files(&project_id, &pattern).await
}

// =============================================================================
// SSE Event Streaming
// =============================================================================

/// Active stream connections - maps stream_id to abort handle
type StreamAbortHandles = Arc<Mutex<HashMap<String, tokio::sync::oneshot::Sender<()>>>>;

/// Get or create the stream abort handles state
fn get_stream_handles(app: &AppHandle) -> StreamAbortHandles {
    if let Some(handles) = app.try_state::<StreamAbortHandles>() {
        handles.inner().clone()
    } else {
        let handles = Arc::new(Mutex::new(HashMap::new()));
        app.manage(handles.clone());
        handles
    }
}

/// Connect to OpenCode SSE event stream
/// 
/// This starts a background task that:
/// 1. Connects to the Management API's SSE endpoint
/// 2. Emits "opencode:event" events to the frontend for each SSE message
/// 3. Emits "opencode:stream-status" events for connection status changes
/// 
/// Returns a StreamConnection with a unique stream_id that can be used to disconnect.
#[tauri::command]
pub async fn opencode_connect_stream(
    app: AppHandle,
    project_id: String,
) -> Result<StreamConnection, AppError> {
    let client = get_client()?;
    let stream_id = uuid::Uuid::new_v4().to_string();
    let stream_id_clone = stream_id.clone();
    let project_id_clone = project_id.clone();
    
    // Create abort channel
    let (abort_tx, mut abort_rx) = tokio::sync::oneshot::channel::<()>();
    
    // Store the abort handle
    let handles = get_stream_handles(&app);
    {
        let mut handles_guard = handles.lock().await;
        handles_guard.insert(stream_id.clone(), abort_tx);
    }
    
    // Get the stream URL and connect
    let response = client.opencode_connect_event_stream(&project_id).await?;
    
    // Spawn background task to process the stream
    let app_clone = app.clone();
    let handles_clone = handles.clone();
    
    tokio::spawn(async move {
        let stream_id = stream_id_clone;
        let project_id = project_id_clone;
        
        // Emit connected status
        let _ = app_clone.emit(
            "opencode:stream-status",
            StreamStatusPayload {
                stream_id: stream_id.clone(),
                project_id: project_id.clone(),
                status: StreamStatus::Connected,
                error: None,
            },
        );
        
        // Process the stream
        let mut bytes_stream = response.bytes_stream();
        let mut buffer = String::new();
        let mut chunk_count = 0u64;
        let mut event_count = 0u64;
        
        tracing::info!("Starting SSE stream processing for project {}", project_id);
        
        loop {
            tokio::select! {
                // Check for abort signal
                _ = &mut abort_rx => {
                    tracing::info!("SSE stream aborted for project {}", project_id);
                    break;
                }
                
                // Process stream data
                chunk = bytes_stream.next() => {
                    match chunk {
                        Some(Ok(bytes)) => {
                            chunk_count += 1;
                            let text = String::from_utf8_lossy(&bytes);
                            tracing::debug!("SSE chunk #{} ({} bytes): {:?}", chunk_count, bytes.len(), 
                                if text.len() > 200 { format!("{}...", &text[..200]) } else { text.to_string() });
                            buffer.push_str(&text);
                            
                            // Process complete SSE events from buffer
                            while let Some(event) = parse_sse_event(&mut buffer) {
                                event_count += 1;
                                tracing::info!("Emitting SSE event #{}: type={}", event_count, event.event_type);
                                let emit_result = app_clone.emit(
                                    "opencode:event",
                                    StreamEventPayload {
                                        stream_id: stream_id.clone(),
                                        project_id: project_id.clone(),
                                        event,
                                    },
                                );
                                if let Err(e) = emit_result {
                                    tracing::error!("Failed to emit SSE event: {}", e);
                                }
                            }
                        }
                        Some(Err(e)) => {
                            tracing::error!("SSE stream error for project {}: {}", project_id, e);
                            // Emit error status
                            let _ = app_clone.emit(
                                "opencode:stream-status",
                                StreamStatusPayload {
                                    stream_id: stream_id.clone(),
                                    project_id: project_id.clone(),
                                    status: StreamStatus::Error,
                                    error: Some(e.to_string()),
                                },
                            );
                            break;
                        }
                        None => {
                            tracing::info!("SSE stream ended for project {} (chunks={}, events={})", 
                                project_id, chunk_count, event_count);
                            break;
                        }
                    }
                }
            }
        }
        
        // Clean up and emit disconnected status
        {
            let mut handles_guard = handles_clone.lock().await;
            handles_guard.remove(&stream_id);
        }
        
        let _ = app_clone.emit(
            "opencode:stream-status",
            StreamStatusPayload {
                stream_id: stream_id.clone(),
                project_id: project_id.clone(),
                status: StreamStatus::Disconnected,
                error: None,
            },
        );
    });
    
    Ok(StreamConnection {
        stream_id,
        project_id,
    })
}

/// Disconnect from an OpenCode SSE event stream
#[tauri::command]
pub async fn opencode_disconnect_stream(
    app: AppHandle,
    stream_id: String,
) -> Result<(), AppError> {
    let handles = get_stream_handles(&app);
    let mut handles_guard = handles.lock().await;
    
    if let Some(abort_tx) = handles_guard.remove(&stream_id) {
        // Send abort signal (ignore error if receiver is already dropped)
        let _ = abort_tx.send(());
    }
    
    Ok(())
}

/// Parse a single SSE event from the buffer, removing it from the buffer if found
fn parse_sse_event(buffer: &mut String) -> Option<OpenCodeEvent> {
    // SSE events are separated by double newlines
    if let Some(event_end) = buffer.find("\n\n") {
        let event_str = buffer[..event_end].to_string();
        buffer.drain(..event_end + 2);
        
        // Skip SSE comments (lines starting with ':')
        // The server sends ": connected" as a keep-alive/connection confirmation
        let is_comment = event_str.lines().all(|line| line.starts_with(':') || line.is_empty());
        if is_comment {
            tracing::debug!("Skipping SSE comment: {:?}", event_str);
            return None;
        }
        
        // Parse the SSE event
        let mut event_type = String::new();
        let mut data = String::new();
        
        for line in event_str.lines() {
            if let Some(value) = line.strip_prefix("event:") {
                event_type = value.trim().to_string();
            } else if let Some(value) = line.strip_prefix("data:") {
                if !data.is_empty() {
                    data.push('\n');
                }
                data.push_str(value.trim());
            }
            // Ignore comment lines starting with ':'
        }
        
        // If no data, skip this event
        if data.is_empty() {
            tracing::debug!("Skipping SSE event with no data: {:?}", event_str);
            return None;
        }
        
        // Store data length for logging before it might be moved
        let data_len = data.len();
        
        // Try to parse data as JSON
        let data_value = serde_json::from_str(&data).unwrap_or_else(|e| {
            tracing::warn!("Failed to parse SSE data as JSON: {} - data: {:?}", e, &data);
            serde_json::Value::String(data)
        });
        
        // If no event type specified via `event:` line, try to extract from JSON data
        // OpenCode sends events like: data: {"type":"message.part.updated","properties":{...}}
        if event_type.is_empty() {
            if let Some(obj) = data_value.as_object() {
                if let Some(type_val) = obj.get("type") {
                    if let Some(t) = type_val.as_str() {
                        event_type = t.to_string();
                    }
                }
            }
        }
        
        // Default to "message" if still no event type
        if event_type.is_empty() {
            event_type = "message".to_string();
        }
        
        tracing::debug!("Parsed SSE event: type={}, data_len={}", event_type, data_len);
        
        Some(OpenCodeEvent {
            event_type,
            data: data_value,
        })
    } else {
        None
    }
}
